File: website/fe_supplier/app/service/reviewService.ts
import api from '~/config/axios';

export interface CreateReviewRequest {
  orderDetailId: string;
  rating: number;
  comment?: string;
  imageUrl?: string;
}

export interface ReplyReviewRequest {
  reply: string;
}

export interface ReviewResponse {
  reviewId: string;
  customerId: string;
  customerName: string;
  productVariantId: string;
  productVariantName: string;
  productId: string;
  productName: string;
  productImage?: string;
  storeId: string;
  storeName: string;
  orderDetailId?: string;
  orderCode?: string;
  rating: number;
  comment?: string;
  imageUrl?: string;
  supplierReply?: string;
  repliedAt?: string;
  markedAsSpam: boolean;
  createdAt: string;
  canEdit: boolean;
  canDelete: boolean;
  canReply: boolean;
  canEditReply: boolean;
}

export interface ProductRatingResponse {
  productId: string;
  productName: string;
  averageRating: number;
  totalReviews: number;
  ratingDistribution: { [key: number]: number };
  ratingPercentage: { [key: number]: number };
}

export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

export const reviewService = {
  // Get reviews for supplier's store
  getStoreReviews: async (storeId: string, page: number = 0, size: number = 10): Promise<PageResponse<ReviewResponse>> => {
    const response = await api.get(`/reviews/store/${storeId}`, {
      params: { page, size }
    });
    return response.data;
  },

  // Get review by ID
  getReviewById: async (reviewId: string): Promise<ReviewResponse> => {
    const response = await api.get(`/reviews/${reviewId}`);
    return response.data;
  },

  // Get product variant reviews
  getProductReviews: async (productVariantId: string, rating?: number, page: number = 0, size: number = 10): Promise<PageResponse<ReviewResponse>> => {
    const response = await api.get(`/reviews/product/${productVariantId}`, {
      params: { rating, page, size }
    });
    return response.data;
  },

  // Get product variant rating statistics
  getProductRating: async (productVariantId: string): Promise<ProductRatingResponse> => {
    const response = await api.get(`/reviews/product/${productVariantId}/rating`);
    return response.data;
  },

  // Search reviews by keyword for product variant
  searchReviews: async (productVariantId: string, keyword: string, page: number = 0, size: number = 10): Promise<PageResponse<ReviewResponse>> => {
    const response = await api.get(`/reviews/product/${productVariantId}/search`, {
      params: { keyword, page, size }
    });
    return response.data;
  },

  // Supplier replies to a review
  replyToReview: async (reviewId: string, reply: string): Promise<ReviewResponse> => {
    const response = await api.post(`/reviews/${reviewId}/reply`, { reply });
    return response.data;
  },

  // Supplier updates their reply
  updateReply: async (reviewId: string, reply: string): Promise<ReviewResponse> => {
    const response = await api.put(`/reviews/${reviewId}/reply`, { reply });
    return response.data;
  },

  // Supplier deletes their reply
  deleteReply: async (reviewId: string): Promise<void> => {
    await api.delete(`/reviews/${reviewId}/reply`);
  },

  // Supplier reports an inappropriate review
  reportReview: async (reviewId: string, reason: string): Promise<void> => {
    await api.post(`/reviews/${reviewId}/report`, null, {
      params: { reason }
    });
  },
};

File: backend/src/main/java/com/example/backend/service/ReviewService.java
package com.example.backend.service;

import com.example.backend.dto.request.CreateReviewRequest;
import com.example.backend.dto.request.UpdateReviewRequest;
import com.example.backend.dto.response.ProductRatingResponse;
import com.example.backend.dto.response.ReviewResponse;
import org.springframework.data.domain.Page;

public interface ReviewService {

    /**
     * Customer creates a review for an order detail
     */
    ReviewResponse createReview(String customerId, CreateReviewRequest request);

    /**
     * Customer updates their review
     */
    ReviewResponse updateReview(String customerId, String reviewId, UpdateReviewRequest request);

    /**
     * Customer deletes their review
     */
    void deleteReview(String customerId, String reviewId);

    /**
     * Get review by ID
     */
    ReviewResponse getReviewById(String reviewId);

    /**
     * Get all reviews by customer
     */
    Page<ReviewResponse> getCustomerReviews(String customerId, int page, int size);

    /**
     * Get all reviews for a product
     */
    Page<ReviewResponse> getProductReviews(String productId, Integer rating, int page, int size);

    /**
     * Get all reviews for a store
     */
    Page<ReviewResponse> getStoreReviews(String storeId, int page, int size);

    /**
     * Get product rating statistics
     */
    ProductRatingResponse getProductRating(String productId);

    /**
     * Admin marks review as spam
     */
    ReviewResponse markAsSpam(String reviewId, boolean isSpam);

    /**
     * Get spam reviews for moderation (admin only)
     */
    Page<ReviewResponse> getSpamReviews(int page, int size);

    /**
     * Search reviews by keyword
     */
    Page<ReviewResponse> searchReviews(String productId, String keyword, int page, int size);

    /**
     * Supplier replies to a review
     */
    ReviewResponse replyToReview(String supplierId, String reviewId, String reply);

    /**
     * Supplier updates their reply
     */
    ReviewResponse updateReply(String supplierId, String reviewId, String reply);

    /**
     * Supplier deletes their reply
     */
    void deleteReply(String supplierId, String reviewId);

    /**
     * Supplier reports a review as inappropriate
     */
    void reportReview(String supplierId, String reviewId, String reason);
}
